src/coord/axisTickLabelBuilder.ts

```diff
@@ -191,95 +233,309 @@ function makeRealNumberLabels(axis: Axis) {
+    return applyPadding({ width: maxW, height: maxH }, labelPadding);
+}
+
+
+/**
+ * Calculate interval for category axis ticks and labels.
+ * To get precise result, at least one of `getRotate` and `isHorizontal`
+ * should be implemented in axis.
+ */
+export function calculateCategoryInterval(axis: Axis) {
+    return calculateCategoryAutoLayout(axis).interval;
+}
+
+
+function layoutScaleStep(tickCount: number, labelModel: Model<AxisBaseOption['axisLabel']>, interval?: number) {
+    // Simple optimization for large amount of labels: trigger sparse label iteration
+    // if tick count is over the threshold
+    const treshhold = labelModel.get('layoutApproximationThreshold') ?? 40;
+    const step = Math.max((interval ?? 0) + 1, 1);
+    return tickCount > treshhold ? Math.max(step, Math.floor(tickCount / treshhold)) : step;
+}
+
+function calculateLabelInterval(unitSize: Size, maxLabelSize: Size, minDistance: number) {
+    let dw = (maxLabelSize.width + minDistance) / unitSize.width;
+    let dh = (maxLabelSize.height + minDistance) / unitSize.height;
     // 0/0 is NaN, 1/0 is Infinity.
     isNaN(dw) && (dw = Infinity);
     isNaN(dh) && (dh = Infinity);
-    let interval = Math.max(0, Math.floor(Math.min(dw, dh)));
 
-    const cache = inner(axis.model);
+    return Math.max(0, Math.floor(Math.min(dw, dh)));
+}
+
+/**
+ * Rotate label's rectangle, see https://codepen.io/agurtovoy/pen/WNPyqWx for the visualization of the math
+ * below.
+ *
+ * @return {Object} {
+ *   axesIntersection: Size, // intersection of the rotated label's rectangle with the corresponding axes
+ *   bounds: Size // dimensions of the rotated label's bounding rectangle
+ *   offset: PointLike // bounding rectangle's offset from the assumed rotation origin
+ *  }
+ */
+function rotateLabel({ width, height }: Size, rotation: number) {
+    const rad = rotation * RADIAN;
+    const sin = Math.abs(Math.sin(rad));
+    const cos = Math.abs(Math.cos(rad));
+
+    // width and height of the intersection of the rotated label's rectangle with the corresponding axes, see
+    // https://math.stackexchange.com/questions/1449352/intersection-between-a-side-of-rotated-rectangle-and-axis
+    const axesIntersection = {
+        width: Math.min(width / cos, height / sin),
+        height: Math.min(height / cos, width / sin)
+    };
+
+    // width and height of the rotated label's bounding rectangle; note th
+    const bounds = {
+        width: width * cos + height * sin,
+        height: width * sin + height * cos
+    };
+
+    // label's bounding box's rotation origin is at the vertical center of the bbox's axis edge
+    // rather than the corresponding corner point
+    const asbRotation = Math.abs(rotation);
+    const bboxOffset = asbRotation === 0 || asbRotation === 180 ? 0 : height / 2;
+    const offset = {
+        x: bboxOffset * sin,
+        y: bboxOffset * cos
+    };
+
+    return { axesIntersection, bounds, offset };
```

Comment:
See https://codepen.io/agurtovoy/pen/WNPyqWx for the visualization of the math here.

<img width="640" alt="label rotation visualization" src="https://github.com/apache/echarts/assets/3394202/bd699411-4459-47b2-af4f-31a11132fec7">